### ISSUE TITLE
Investgate DDP model/gradient sync

### ISSUE THREAD

**Issue by jkelling** [Photon-AI-Research/InSituML#122]
In data parallel training, bad and varying learning curves were observed. Possible for this need to be investigated:

- [x] Are the model weights correctly synchronized upon initialization/loading?
- [x] Are the obtained model weights on all ranks identical (i.e. are the gradients synced correctly?)
- [x] Are loss calculations correct for distributed batches?
  - [x] KL?
    - [KL](https://github.com/Photon-AI-Research/InSituML/blob/khi/main/ModelHelpers/cINN/train_khi_AE_refactored/utilities.py#L94-L95) is probably save as it is computed element-wise.
  - [x] MMD?
    - [MMD](https://github.com/Photon-AI-Research/InSituML/blob/khi/main/ModelHelpers/cINN/utilities.py#L10) may be wrong in that with DDP each generated point is not compared to *all* samples in the batch, only the samples local to the rank. Solution may be to [all_gather](https://pytorch.org/docs/stable/distributed.html#torch.distributed.all_gather) the data samples to compute blocks of rows/columns of the total matrix locally instead of only diagonal blocks. The averaging of gradients may be enough to take care of filling the other matrix axis.
